### PR TITLE
feature: support images filter flag

### LINKED ISSUE #2371: [question] Is there a need to support pouch images filter and no-trunc flag?
### Ⅰ. Issue Description
As the title desc.

### Ⅱ. Describe what happened
None.

### Ⅲ. Describe what you expected to happen
None.

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1.
2.
3.

### Ⅴ. Anything else we need to know?
None.

### Ⅵ. Environment:
None.

- pouch version (use `pouch version`):
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- `Others:```

### PR DESCRIPTION
**Pull request by ZYecho**
Signed-off-by: zhangyue <zy675793960@yeah.net>

### Ⅰ. Describe what this PR did
add filter flag support for `pouch images`, support three filters: since/before/reference

- before (<image-name>[:<tag>], <image id> or <image@digest>) - filter images created before given id or references
- since (<image-name>[:<tag>], <image id> or <image@digest>) - filter images created since given id or references
- The reference filter shows only images whose reference matches the specified pattern. may have multi reference filter. the relationship between multi reference filter is or.

### Ⅱ. Does this pull request fix one issue?

Related to #2405 
fix #2371 

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)
add api testcase and cli testcase.

### Ⅳ. Describe how to verify it
`pouch images`
```
IMAGE ID       IMAGE NAME                                          SIZE
8c811b4aec35   registry.hub.docker.com/library/busybox:1.28        710.81 KB
4ab4c602aa5e   registry.hub.docker.com/library/hello-world:linux   5.25 KB
```

`pouch images -f reference=registry.hub.docker.com/library/busybox:*`
```
IMAGE ID       IMAGE NAME                                     SIZE
8c811b4aec35   registry.hub.docker.com/library/busybox:1.28   710.81 KB
```

`pouch images -f after=registry.hub.docker.com/library/busybox:1.28`
Got
`Error: failed to get image list: {"message":"Invalid filter after"}`

`pouch images -f before=registry.hub.docker.com/library/hello-world:linux`
```
IMAGE ID       IMAGE NAME                                     SIZE
8c811b4aec35   registry.hub.docker.com/library/busybox:1.28   710.81 KB
```

### Ⅴ. Special notes for reviews
Use the same filter package as `pouch events`

### CODE REVIEW

**Comment by allencloud** [comment]
CI fails with code check. Please take a look. @ZYecho Thanks.

**Comment by ZYecho** [comment]
> CI fails with code check. Please take a look. @ZYecho Thanks.

@allencloud  fixed it,  also cc @Ace-Tang @fuweid request for review.

File: apis/filters/parse.go [hunk 1]
```diff
@@ -150,3 +157,33 @@ func FromParam(p string) (Args, error) {
 	}
 	return args, nil
 }
+
+//FromFilterOpts parse key=value to Args string from cli opts
```
Comment (HusterWan) [inline apis/filters/parse.go]: Maybe we can add some test cases for this new function, WDYT ? @ZYecho

File: daemon/mgr/image.go [hunk 2]
```diff
@@ -21,14 +22,20 @@ import (
...
-	digest "github.com/opencontainers/go-digest"
+	"github.com/opencontainers/go-digest"
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 	pkgerrors "github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 )
 
 var deadlineLoadImagesAtBootup = time.Second * 10
 
+var acceptedImageFilterTags = map[string]bool{
```
Comment (HusterWan) [inline daemon/mgr/image.go]: explain this tags?

File: apis/filters/parse.go (hunk 1 above)
Comment (ZYecho) [reply apis/filters/parse.go]: good idea, I add a testcase to cover this.

File: daemon/mgr/image.go (hunk 2 above)
Comment (ZYecho) [reply daemon/mgr/image.go]: done.

File: apis/filters/parse.go [hunk 3]
```diff
@@ -150,3 +157,33 @@ func FromParam(p string) (Args, error) {
...
+	}
+	return filterArgs, nil
+}
+
+// Validate compared the set of accepted keys against the keys in the mapping.
+// An error is returned if any mapping keys are not in the accepted set.
+func (args Args) Validate(accepted map[string]bool) error {
+	for name := range args.fields {
+		if !accepted[name] {
+			return errors.New("Invalid filter " + name)
```
Comment (fuweid) [inline apis/filters/parse.go]: could we avoid to start with a capital letter in the error string? The reason is here https://github.com/golang/go/wiki/Errors#errors

File: daemon/mgr/image.go [hunk 4]
```diff
@@ -161,12 +169,80 @@ func (mgr *ImageManager) GetImage(ctx context.Context, idOrRef string) (*types.I
...
+		if sinceFilter != nil {
+			if img.OCISpec.Created.Equal(sinceTime) || img.OCISpec.Created.Before(sinceTime) {
+				continue
+			}
+		}
+
+		if filter.Contains("reference") {
+			var found bool
+			for _, ref := range mgr.localStore.GetReferences(img.ID) {
+				for _, pattern := range filter.Get("reference") {
```
Comment (fuweid) [inline daemon/mgr/image.go]: could we retrieve `filter.Get("reference")` before? I think the `Get` will allocate new memory each loop.

File: apis/filters/parse.go (hunk 3 above)
Comment (Ace-Tang) [reply apis/filters/parse.go]: could you add accepted filter name ? make user more clearly

File: apis/server/image_bridge.go [hunk 5]
```diff
@@ -75,9 +76,12 @@ func (s *Server) getImage(ctx context.Context, rw http.ResponseWriter, req *http
 }
 
 func (s *Server) listImages(ctx context.Context, rw http.ResponseWriter, req *http.Request) error {
-	filters := req.FormValue("filters")
+	filter, err := filters.FromParam(req.FormValue("filters"))
```
Comment (Ace-Tang) [inline apis/server/image_bridge.go]: any reason to change `filters`  to `filter` ?

File: cli/image_list.go [hunk 6]
```diff
@@ -4,11 +4,12 @@ import (
 	"context"
 	"fmt"
 
+	"github.com/alibaba/pouch/apis/filters"
 	"github.com/alibaba/pouch/apis/types"
 	"github.com/alibaba/pouch/pkg/reference"
 	"github.com/alibaba/pouch/pkg/utils"
 
-	digest "github.com/opencontainers/go-digest"
```
Comment (Ace-Tang) [inline cli/image_list.go]: Is this related to goimport version ? @fuweid @HusterWan

File: daemon/mgr/image.go (hunk 4 above)
Comment (Ace-Tang) [reply daemon/mgr/image.go]: seems much condition here

File: cli/image_list.go (hunk 6 above)
Comment (fuweid) [reply cli/image_list.go]: so far it is ok.

File: daemon/mgr/image.go (hunk 4 above)
Comment (fuweid) [reply daemon/mgr/image.go]: what is the condition?

File: apis/filters/parse.go (hunk 3 above)
Comment (ZYecho) [reply apis/filters/parse.go]: > could you add accepted filter name ? make user more clearly

Is there a need to do here?  I find that other validate func in pouch also not do this, maybe We can do this by adding docs or cli usage? wdyt? @Ace-Tang

File: apis/filters/parse_test.go [hunk 7]
```diff
@@ -149,3 +149,30 @@ func TestFromParam(t *testing.T) {
 		}
 	}
 }
+
+func TestFromFilterOpts(t *testing.T) {
+	filterOpts := []string{
+		"reference=img1",
```
[the remark this answers is not in the document]
Comment (ZYecho) [reply apis/filters/parse_test.go]: now not support comma like reference=img1, img2.

File: daemon/mgr/image.go [hunk 8]
```diff
@@ -161,12 +169,80 @@ func (mgr *ImageManager) GetImage(ctx context.Context, idOrRef string) (*types.I
 }
 
 // ListImages lists images stored by containerd.
-func (mgr *ImageManager) ListImages(ctx context.Context, filter ...string) ([]types.ImageInfo, error) {
-	// TODO: support filter functionality
+func (mgr *ImageManager) ListImages(ctx context.Context, filter filters.Args) ([]types.ImageInfo, error) {
 	ctrdImageInfos := mgr.localStore.ListCtrdImageInfo()
 	imgInfos := make([]types.ImageInfo, 0, len(ctrdImageInfos))
 
+	if err := filter.Validate(acceptedImageFilterTags); err != nil {
```
[the remark this answers is not in the document]
Comment (ZYecho) [reply daemon/mgr/image.go]: updated.

File: apis/server/image_bridge.go (hunk 5 above)
Comment (fuweid) [reply apis/server/image_bridge.go]: how about this? change filters into apifilters? just ask.

File: daemon/mgr/image.go [hunk 9]
```diff
@@ -161,12 +169,80 @@ func (mgr *ImageManager) GetImage(ctx context.Context, idOrRef string) (*types.I
...
+	var (
+		beforeFilter, sinceFilter *types.ImageInfo
+		beforeTime, sinceTime     time.Time
+		err                       error
+	)
+
+	beforeImages := filter.Get("before")
+	if len(beforeImages) > 0 {
+		// use the last one if multi before images was assigned
+		beforeFilter, err = mgr.GetImage(ctx, beforeImages[len(beforeImages)-1])
```
[the remark this answers is not in the document]
Comment (fuweid) [reply daemon/mgr/image.go]: so, could we limit the multi-since/multi-before here? Since there is no reason for this requirement.

[the remark this answers is not in the document]
Comment (ZYecho) [reply daemon/mgr/image.go]: @fuweid I test the same scenario in moby,  have the same problem.  I think the single-since/since-before is reasonable, while multi is undefined.  So in here, I want to throw err if len > 1, wdyt?

[the remark this answers is not in the document]
Comment (fuweid) [reply daemon/mgr/image.go]: agree

File: test/cli_images_test.go [hunk 10]
```diff
@@ -81,6 +84,36 @@ func (suite *PouchImagesSuite) TestImagesWorks(c *check.C) {
...
+	//Test since filter
+	sinceRes1 := command.PouchRun("images", "-f", "since="+busyboxImage)
+	items1 = imagesListToKV(sinceRes1.Combined())
+	sinceRes2 := command.PouchRun("images", "-f", "since="+helloworldImage)
+	items2 = imagesListToKV(sinceRes2.Combined())
+	c.Assert(len(items1)+len(items2), check.Equals, 1)
+
+	//Test invalid filter
+	invalidRes := command.PouchRun("images", "-f", "after="+busyboxImage)
+	c.Assert(invalidRes.Stderr(), check.Equals, "Error: failed to get image list: {\"message\":\"Invalid filter after\"}\n\n")
```
Comment (fuweid) [inline test/cli_images_test.go]: the case is failed 

```
/home/travis/gopath/src/github.com/alibaba/pouch/test/cli_images_test.go:114:
    c.Assert(invalidRes.Stderr(), check.Equals, "Error: failed to get image list: {\"message\":\"Invalid filter after\"}\n\n")
... obtained string = "" +
...     "Error: failed to get image list: {\"message\":\"invalid filter after\"}\n" +
...     "\n"
... expected string = "" +
...     "Error: failed to get image list: {\"message\":\"Invalid filter after\"}\n" +
...     "\n"

```

I was thinking that we can do this in api level test and only check the status. The message isn't out of control.
But the status should be stable. WDYT?

Comment (ZYecho) [reply test/cli_images_test.go]: agree, update it later.

**Comment by ZYecho** [comment]
@fuweid all mentioned have fixed.  PTAL, thanks.

File: daemon/mgr/image.go [hunk 11]
```diff
@@ -161,12 +169,85 @@ func (mgr *ImageManager) GetImage(ctx context.Context, idOrRef string) (*types.I
...
+	if err := filter.Validate(acceptedImageFilterTags); err != nil {
+		return nil, err
+	}
+
+	beforeImages := filter.Get("before")
+	sinceImages := filter.Get("since")
+
+	// refuse undefined behavior
+	if len(beforeImages) > 1 || len(sinceImages) > 1 {
+		return nil, pkgerrors.Wrapf(errtypes.ErrInvalidParam, "can't use since or before filter more than one")
```
[the remark this answers is not in the document]
Comment (ZYecho) [reply daemon/mgr/image.go]: done.

File: test/api_image_list_test.go [hunk 12]
```diff
@@ -56,6 +59,46 @@ func (suite *APIImageListSuite) TestImageListDigest(c *check.C) {
...
+	c.Assert(got[0].ID, check.NotNil)
+	c.Assert(got[0].CreatedAt, check.NotNil)
+	c.Assert(got[0].Config, check.NotNil)
+	c.Assert(got[0].Architecture, check.NotNil)
+	c.Assert(got[0].Size, check.NotNil)
+	c.Assert(got[0].Os, check.NotNil)
+	c.Assert(reflect.DeepEqual(got[0].RepoTags, []string{repoTag}), check.Equals, true)
+	c.Assert(reflect.DeepEqual(got[0].RepoDigests, []string{repoDigest}), check.Equals, true)
+
+	// check invalid filter
```
[the remark this answers is not in the document]
Comment (ZYecho) [reply test/api_image_list_test.go]: done.

File: client/image_list_test.go [hunk 13]
```diff
@@ -12,14 +12,15 @@ import (
 
 	"github.com/alibaba/pouch/apis/types"
 
+	"github.com/alibaba/pouch/apis/filters"
```
Comment (HusterWan) [inline client/image_list_test.go]: put this line in the second part, thanks @ZYecho

Comment (ZYecho) [reply client/image_list_test.go]: thanks, update.

**Comment by Ace-Tang** [comment]
lgtm

**Comment by HusterWan** [comment]
LGTM, thanks for your works @ZYecho
